WIP: add 0.4.0 blog post (#478)
reflex-dev/reflex-web: blog/2024-02-14-reflex-v0.4.0.md

---
author: Nikhil Rao
date: 2024-02-14
title: "Reflex v0.4.0"
description: "New features coming in the 0.4.0 release."
image: /reflex-040.png
---

```python exec
import reflex as rx

from pcweb.pages.docs import styling, library as lib
from pcweb.pages.docs.library import library
```

We’re excited the announce the next major release of Reflex! 

With Reflex 0.4.0 we’re releasing [new core components]({library.path}) and a [theming system]({styling.theming.path}) to make it easier to build apps that look great out of the box.

## Our Vision

Our goal is for Reflex to be the easiest way to go from an idea to a fully built web app.

Since we launched Reflex, we’ve been working to maintain the balance of having an ergonomic, batteries-included web framework that is also flexible enough to handle any type of use case.

Reflex 0.4.0 takes another step in this direction with our new core components and themes.
The components come with high-level styling options, and different components within an app can be styled together with a unified theme.

> If you want more customizability, you can modify any aspects of the component internals to suit your needs.

Reviewer: ```suggestion
If you want more customizability, you can modify most aspects of the component internals to suit your needs.
```

## Upgrading Reflex

If you’re creating a new project, simply install reflex to automatically use the latest version.

```text
pip install reflex
```

If you have an existing project, upgrade your version of reflex

```text
pip install --upgrade reflex
```

### Keeping Chakra

Since the core components have changed from Radix, existing apps will need to run the migration script to keep using Chakra.

```text
reflex scripts keep-chakra
```

This will convert all components in your app to the new namespace.
For example, if you used an `rx.box` in your app, it will be converted to an `rx.chakra.box`.
You will still be able to use any Chakra component through the `rx.chakra` namespace. 

The [Chakra Docs](https://reflex.dev/docs/library/chakra/datadisplay/badge/) have been moved and updated to reflect the new namespace.

## New Core Components

> We’re switching our core components to be based on the [Radix UI](https://www.radix-ui.com) component library. For example, `rx.button` will now render a Radix button. See the new core component library to see all available components in the core namespace.

Reviewer: ```suggestion
We’re switching our core components to be based on the [Radix UI](https://www.radix-ui.com) component library. For example, `rx.button` will now render a Radix button. Browse the new [core component library]({library.path}) to see all available components in the core namespace.
```

Many components feature high level styling props, such as setting the `color_scheme` or the `variant`. If you need to customize further, you can use any CSS as always.

## Theming System

![Theming Demo](/dashboard.gif)

The new components can be styled together with the new [unified theming system]({styling.theming.path}).

You can specify a theme in your `rx.App` to give your app a unified look and feel such as enabling dark mode, or setting an `accent_color` to match your brand.

```python
app = rx.App(
	theme=rdxt.theme(
	    color_mode="dark", accent_color="blue"
	)
)
```

You can also use the [ThemPanel]({lib.theming.theme_panel.path}) component when developing, to see how your app will look with different theming options.

## Component Namespaces

> Some components are composed of different subcomponents. For example, under the hood an `[rx.select](http://rx.select)` has the following:

Reviewer: ```suggestion
Some components are composed of different subcomponents. For example, under the hood an `rx.select` has the following:
```

- trigger
- content

> For many components, you can use the core namespace to access the high-level API. For example, you can do `rx.select(list)` to render a list of options. If you want further customizability, you can use the low level API. For example, to customize the trigger. Components with a low-level API will have an “Internal” tab on their docs that you can use. Over time, we will add more high-level APIs based on community feedback and usage patterns.

Reviewer: This section should link to the internal page of the rx.select docs so people can get a feel for what you mean without regurgitating code examples in the blog.

## Colors and Color Modes

> This release also features an rx.color module that makes it easier to create colors and shades of colors throughout your app. In the future we will also add ways to define your own color palettes.

Reviewer: ```suggestion
This release also includes the `rx.color` module that makes it easier to create colors and shades of colors throughout your app. In the future we will also add ways to define your own color palettes.
```
